File: _includes/code/python/howto.configure.rbac.oidc.users.py
from weaviate.classes.rbac import Permissions

# START AdminClient
import weaviate
from weaviate.classes.init import Auth

# Connect to Weaviate as root user
client = weaviate.connect_to_local(
    # END AdminClient
    # Use custom port defined in tests/docker-compose-rbac.yml (without showing the user)
    port=8580,
    grpc_port=50551,
    # START AdminClient
    auth_credentials=Auth.api_key("user-a-key"),
)
# END AdminClient

from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.data(collection="TargetCollection*", read=True, create=True),
    Permissions.data(collection="TargetCollection*", read=True, create=False),
]

client.roles.add_permissions(permissions=permissions, role_name="testRole")

# START AssignOidcUserRole
client.users.oidc.assign_roles(user_id="user-b", role_names=["testRole", "viewer"])
# END AssignOidcUserRole
assert "testRole" in client.users.oidc.get_assigned_roles("user-b")
assert "viewer" in client.users.oidc.get_assigned_roles("user-b")

# START ListOidcUserRoles
user_roles = client.users.oidc.get_assigned_roles("user-b")

for role in user_roles:
    print(role)
# END ListOidcUserRoles
assert any(
    permission.collection == "TargetCollection*"
    for permission in user_roles["testRole"].collections_permissions
)
assert any(
    permission.collection == "TargetCollection*"
    for permission in user_roles["testRole"].data_permissions
)

# START RevokeOidcUserRoles
client.users.oidc.revoke_roles(user_id="user-b", role_names="testRole")
# END RevokeOidcUserRoles
assert "testRole" not in client.users.oidc.get_assigned_roles("user-b")


client.close()

File: _includes/code/typescript/howto.configure.rbac.oidc.users.ts
// TODO[g-despot]: OIDC testing not yet implemented
import assert from 'assert'

// START AdminClient
import weaviate, { type WeaviateClient } from 'weaviate-client'

// Connect to Weaviate as root user
const client: WeaviateClient = await weaviate.connectToLocal({
    // END AdminClient
    // Use custom port defined in tests/docker-compose-rbac.yml (without showing the user)
        port: 8580,
        grpcPort: 50551, 
        // START AdminClient
        authCredentials: new weaviate.ApiKey("root-user-key")
})
// END AdminClient

const { permissions } = weaviate

const collectionPermissions = [
    permissions.collections({
        collection: "TargetCollection*", 
        read_config: true, 
        create_collection: true
    }),
    permissions.data({
        collection: "TargetCollection*", 
        read: true, 
        create: true}
        ),
]

await client.roles.delete("testRole")
await client.roles.create("testRole", collectionPermissions)

// START AssignOidcUserRole
await client.users.oidc.assignRoles(["testRole", "viewer"], "custom-user",)
// END AssignOidcUserRole
assert.equal((Object.keys(await client.users.getAssignedRoles("custom-user")).some(
    role => role == "testRole"
)), true)

assert.equal((Object.keys(await client.users.getAssignedRoles("custom-user")).some(
    role => role == "viewer"
)), true)

// START ListOidcUserRoles
const userRoles = await client.users.oidc.getAssignedRoles("custom-user")

for (const [role, value] of Object.entries(userRoles)) {
    console.log(role)
}
// END ListOidcUserRoles
assert.equal((Object.keys(await client.users.db.getAssignedRoles("custom-user")).some(
    role => role == "testRole"
)), true)
assert.equal((Object.keys(await client.users.db.getAssignedRoles("custom-user")).some(
    role => role == "viewer"
)), true)

// START RevokeOidcUserRoles
await client.users.oidc.revokeRoles("testRole","custom-user")
// END RevokeOidcUserRoles
assert.equal((Object.keys(await client.users.db.getAssignedRoles("custom-user")).some(
    role => role == "testRole"
)), false)

client.close()
